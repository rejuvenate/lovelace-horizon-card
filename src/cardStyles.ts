import { css } from 'lit'

export default css`
  .horizon-card {
    --horizon-card-primary: var(--primary-text-color, #000000);
    --horizon-card-secondary: var(--secondary-text-color, #828282);
    --horizon-card-accent: #d7d7d7;

    --horizon-card-lines: var(--horizon-card-accent);
    --horizon-card-field-name-color: var(--horizon-card-secondary);
    --horizon-card-field-value-color: var(--horizon-card-primary);

    --horizon-card-stop-invisible: rgb(0,0,0,0);
    --horizon-card-stop-sun-color: #f9d05e;
    --horizon-card-stop-dawn-color: #393b78;
    --horizon-card-stop-day-color: #8ebeeb;
    --horizon-card-stop-dusk-color: #393b78;

    padding: 0.5rem;
    font-size: 1.3rem;
    font-family: var(--primary-font-family);
  }

  .horizon-card.horizon-card-dark {
    --horizon-card-primary: #ffffff;
    --horizon-card-secondary: #828282;
    --horizon-card-accent: #464646;
  }

  .horizon-card-field-row {
    display: flex;
    justify-content: space-around;
    margin: 1rem 1.5rem 0 1.5rem;
  }

  .horizon-card-text-container {
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  .horizon-card-field-name {
    color: var(--horizon-card-field-name-color);
  }

  .horizon-card-field-value {
    color: var(--horizon-card-field-value-color);
  }

  .horizon-card-header {
    display: flex;
    justify-content: space-between;
    margin: 0 4rem 0 4rem;
  }

  .horizon-card-header .horizon-card-field-value {
    font-size: 1.85rem;
  }

  .horizon-card-title {
    margin: 0.5rem 0.5rem 3rem 3rem;
    font-size: 2rem;
    font-weight: 500;
    color: var(--horizon-card-primary);
  }

  .horizon-card-graph {
    shape-rendering="geometricPrecision";
    margin: 2rem 0 2rem 0;
  }

  .horizon-card-graph .sunInitialStop {
    stop-color: var(--horizon-card-stop-sun-color);
  }

  .horizon-card-graph .sunMiddleStop {
    stop-color: var(--horizon-card-stop-sun-color);
  }

  .horizon-card-graph .sunEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .horizon-card-graph .dawnInitialStop {
    stop-color: var(--horizon-card-stop-dawn-color);
  }

  .horizon-card-graph .dawnMiddleStop {
    stop-color: var(--horizon-card-stop-dawn-color);
  }

  .horizon-card-graph .dawnEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .horizon-card-graph .dayInitialStop {
    stop-color: var(--horizon-card-stop-day-color);
  }

  .horizon-card-graph .dayMiddleStop {
    stop-color: var(--horizon-card-stop-day-color);
  }

  .horizon-card-graph .dayEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .horizon-card-graph .duskInitialStop {
    stop-color: var(--horizon-card-stop-dusk-color);
  }

  .horizon-card-graph .duskMiddleStop {
    stop-color: var(--horizon-card-stop-dusk-color);
  }

  .horizon-card-graph .duskEndStop {
    stop-color: var(--horizon-card-stop-invisible);
  }

  .card-config ul {
    list-style: none;
    padding: 0 0 0 1.5rem;
  }

  .card-config li {
    padding: 0.5rem 0;
  }
`
