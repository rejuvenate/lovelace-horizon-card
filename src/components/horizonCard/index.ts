export * from './HorizonCard'
export * from './HorizonCardContent'
export * from './HorizonCardFooter'
export * from './HorizonCardGraph'
export * from './HorizonCardHeader'
