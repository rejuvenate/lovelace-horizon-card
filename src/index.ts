import './components/horizonCard'
import './components/horizonCardEditor'

