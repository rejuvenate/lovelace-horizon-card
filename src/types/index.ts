import { LovelaceCardConfig } from 'custom-card-helpers'

import { I18N } from '../utils/I18N'

export type THorizonCardFields = {
  sunrise?: boolean
  sunset?: boolean

  dawn?: boolean
  noon?: boolean
  dusk?: boolean

  azimuth?: boolean
  elevation?: boolean
}

export interface IHorizonCardConfig extends LovelaceCardConfig {
  i18n?: I18N
  darkMode?: boolean
  language?: string

  title?: string,
  component?: string
  use12hourClock?: boolean

  fields?: THorizonCardFields
}

export interface IConfigChangedEvent <T> extends CustomEvent<T> {
  target: CustomEvent['target'] & {
    configValue: string
    selected?: string
    checked?: boolean
  }
}

export type TSunInfo = {
  sunrise: number,
  sunset: number

  dawnProgressPercent: number
  dayProgressPercent: number
  duskProgressPercent: number

  sunAboveHorizon: boolean
  sunPercentOverHorizon: number
  sunPosition: {
    x: number
    y: number
  }
}

export enum EHorizonCardErrors {
  SunIntegrationNotFound = 'SunIntegrationNotFound'
}

export type THorizonCardTimes = {
  dawn?: Date
  dusk?: Date
  noon: Date
  sunrise?: Date
  sunset?: Date
}

export type THorizonCardData = {
  azimuth: number
  elevation: number

  sunInfo: TSunInfo

  times: THorizonCardTimes

  error?: EHorizonCardErrors
}

export enum EHorizonCardI18NKeys {
  Azimuth = 'azimuth',
  Dawn = 'dawn',
  Dusk = 'dusk',
  Elevation = 'elevation',
  Noon = 'noon',
  Sunrise = 'sunrise',
  Sunset = 'sunset'
}

export type THorizonCardI18N = Record<string, unknown>

export type THorizonCardI18NErrorKeys = {
  [key in EHorizonCardErrors]: string
}

export type THorizonCardI18NKeys = { [key in EHorizonCardI18NKeys ]: string } | { errors: THorizonCardI18NErrorKeys }
