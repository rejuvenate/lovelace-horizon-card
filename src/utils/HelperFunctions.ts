import { html, nothing, TemplateResult } from 'lit'

import { Constants } from '../constants'
import { EHorizonCardI18NKeys, TMoonPhase } from '../types'
import { I18N } from './I18N'

type FieldValue = Date | number | string | undefined

export class HelperFunctions {
  public static renderFieldElements (i18n: I18N, translationKey: string, values: FieldValue[],
                                     extraClasses: string[] = []): TemplateResult {
    const mappedValues = values
      .map((value, index) => this.valueToHtml(i18n, translationKey, value, extraClasses[index]))

    return this.renderFieldElement(i18n, translationKey, mappedValues)
  }

  public static renderFieldElement (i18n: I18N, translationKey: string, value: FieldValue | TemplateResult[]): TemplateResult {
    return html`
      <div class="horizon-card-text-container">
        <div class="horizon-card-field-name">${ i18n.tr(translationKey) }</div>
        ${value instanceof Array ? value : this.valueToHtml(i18n, translationKey, value) }
      </div>
    `
  }

  public static renderMoonElement (i18n: I18N, phase: TMoonPhase, phaseRotation: number) {
    if (phase === undefined) {
      return nothing
    }

    let moon_phase_localized: unknown = i18n.localize(`component.moon.entity.sensor.phase.state.${phase.state}`)
    if (!moon_phase_localized) {
      moon_phase_localized = html`<span title="Install Moon integration to get localized Moon phase name">${phase.state} (!)</span>`
    }

    return html`
      <div class="horizon-card-text-container">
        <div class="horizon-card-field-moon-phase" style="transform: rotate(${phaseRotation}deg)">
          <ha-icon icon="mdi:${phase.icon}"></ha-icon>
        </div>
        <div class="horizon-card-field-value horizon-card-field-value-moon-phase">${moon_phase_localized}</div>
      </div>
    `
  }

  private static valueToHtml (i18n: I18N, translationKey: string, value: FieldValue, klass='') {
    const mappedValue = this.fieldValueToString(i18n, translationKey, value)
    return html`<div class="horizon-card-field-value ${klass}">${mappedValue}</div>`
  }

  private static fieldValueToString (i18n: I18N, translationKey: string, value: FieldValue) {
    let pre = ''
    let post = ''
    if (value === undefined) {
      value = '-'
    } else if (value instanceof Date) {
      value = i18n.formatDateAsTime(value)
      const parts = value.match(/(.*?)(\d{1,2}[:.]\d{2})(.*)/)
      if (parts != null) {
        pre = parts[1]
        value = parts[2]
        post = parts[3]
      }
    } else if (typeof value === 'number') {
      value = i18n.formatDecimal(value)
      if (translationKey === EHorizonCardI18NKeys.Azimuth || translationKey === EHorizonCardI18NKeys.Elevation) {
        value += 'Â°'
      }
    }
    const preHtml = pre ? html`<span class="horizon-card-field-value-secondary">${pre}</span>` : nothing
    const postHtml = post ? html`<span class="horizon-card-field-value-secondary">${post}</span>` : nothing

    return html`${preHtml}${value}${postHtml}`
  }

  public static isValidLanguage (language: string): boolean {
    return Object.keys(Constants.LOCALIZATION_LANGUAGES).includes(language)
  }

  public static clamp (min: number, max: number, value: number): number {
    if (min === max) {
      return min
    }

    if (min > max) {
      throw new RangeError('Min value can not be bigger than the max value')
    }

    return Math.min(Math.max(value, min), max)
  }

  public static rangeScale (minRange: number, maxRange: number, range: number, value: number) {
    const clamped = HelperFunctions.clamp(minRange, maxRange, range) - minRange
    const rangeSize = maxRange - minRange
    return (1 - clamped / rangeSize) * value
  }

  public static noonAtTimeZone (date: Date, timeZone: string): Date {
    let tzDate
    try {
      tzDate = this.getTimeInTimeZone(date, '12:00:00', timeZone)
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(e)
      tzDate = new Date(date)
      tzDate.setHours(12)
      tzDate.setMinutes(0)
      tzDate.setSeconds(0)
      tzDate.setMilliseconds(0)
    }
    return tzDate
  }

  public static midnightAtTimeZone (date: Date, timeZone: string): Date {
    let tzDate
    try {
      tzDate = this.getTimeInTimeZone(date, '00:00:00', timeZone)
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(e)
      tzDate = new Date(date)
      tzDate.setHours(0)
      tzDate.setMinutes(0)
      tzDate.setSeconds(0)
      tzDate.setMilliseconds(0)
    }
    return tzDate
  }

  private static getTimeInTimeZone (date: Date, time: string, timeZone: string) {
    const formatter = new Intl.DateTimeFormat('fr-CA',
      { timeZone: timeZone, timeZoneName: 'longOffset' })
    // 'fr-CA' locale formats like '2023-04-11 UTC+03:00' or '2023-04-11 UTC-10:00' or '2023-04-11 UTC'
    const formatted = formatter.format(date)
    const parts = formatted
      .replace('\u2212', '-') // minuses might be U+2212 instead of plain old ASCII hyphen-minus
      .split(' ')
    let tz = parts[1].replace('UTC', '')
    if (tz === '') {
      tz = 'Z'
    }
    const dateToParse = `${parts[0]}T${time}${tz}`
    const result = new Date(dateToParse)
    if (isNaN(result.getTime())) {
      // Something went fishy with using the above method - generally should not happen
      throw new Error(`Could not convert time to time zone: ${formatted} -> ${dateToParse}`)
    }
    return result
  }
}
