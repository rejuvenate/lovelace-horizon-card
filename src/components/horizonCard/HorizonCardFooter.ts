import { html, TemplateResult } from 'lit'

import { EHorizonCardI18NKeys, IHorizonCardConfig, THorizonCardData, THorizonCardFields, THorizonCardTimes } from '../../types'
import { HelperFunctions } from '../../utils/HelperFunctions'
import { I18N } from '../../utils/I18N'

export class HorizonCardFooter {
  private data: THorizonCardData
  private i18n: I18N
  private times: THorizonCardTimes
  private fields: THorizonCardFields

  constructor (config: IHorizonCardConfig, data: THorizonCardData) {
    this.data = data

    this.i18n = config.i18n!
    this.times = data?.times
    this.fields = config.fields!
  }

  public render (): TemplateResult {
    return html`
      <div class="horizon-card-footer">
        <div class="horizon-card-field-row">
          ${
  this.fields?.dawn !== undefined
    ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Dawn, this.times?.dawn)
    : HelperFunctions.nothing()
}
          ${
  this.fields?.noon !== undefined
    ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Noon, this.times?.noon)
    : HelperFunctions.nothing()
}
          ${
  this.fields?.dusk !== undefined
    ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Dusk, this.times?.dusk)
    : HelperFunctions.nothing()
}
        </div>

        <div class="horizon-card-field-row">
          ${
  this.fields?.azimuth !== undefined
    ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Azimuth, this.data?.azimuth)
    : HelperFunctions.nothing()
}
          ${
  this.fields?.elevation !== undefined
    ? HelperFunctions.renderFieldElement(this.i18n, EHorizonCardI18NKeys.Elevation, this.data?.elevation)
    : HelperFunctions.nothing()
}
        </div>
      </div>
    `
  }
}
