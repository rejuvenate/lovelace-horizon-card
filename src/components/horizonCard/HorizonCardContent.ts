import { html,TemplateResult } from 'lit'

import { IHorizonCardConfig,THorizonCardData } from '../../types'
import { HelperFunctions } from '../../utils/HelperFunctions'
import { HorizonCardFooter } from './HorizonCardFooter'
import { HorizonCardGraph } from './HorizonCardGraph'
import { HorizonCardHeader } from './HorizonCardHeader'

export class HorizonCardContent {
  private config: IHorizonCardConfig
  private data: THorizonCardData

  constructor (config: IHorizonCardConfig, data: THorizonCardData) {
    this.config = config
    this.data = data
  }

  render (): TemplateResult {
    return html`
      <ha-card>
        <div class="horizon-card ${this.config.darkMode ? 'horizon-card-dark' : ''}">
          ${ this.showHeader() ? this.renderHeader() : HelperFunctions.nothing() }
          ${ this.renderGraph() }
          ${ this.showFooter() ? this.renderFooter() : HelperFunctions.nothing() }
        </div>
      </ha-card>
    `
  }

  private renderHeader (): TemplateResult {
    return new HorizonCardHeader(this.config, this.data).render()
  }

  private renderGraph (): TemplateResult {
    return new HorizonCardGraph(this.data).render()
  }

  private renderFooter (): TemplateResult {
    return new HorizonCardFooter(this.config, this.data).render()
  }

  private showHeader (): boolean {
    // logic based on config
    return true
  }

  private showFooter (): boolean {
    // logic based on config
    return true
  }
}
