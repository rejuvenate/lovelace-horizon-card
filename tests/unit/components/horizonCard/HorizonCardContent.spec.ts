import { html } from 'lit'

import { HorizonCardContent } from '../../../../src/components/horizonCard'
import { IHorizonCardConfig, THorizonCardData } from '../../../../src/types'
import { HelperFunctions } from '../../../../src/utils/HelperFunctions'
import { CustomSnapshotSerializer, TemplateResultTestHelper } from '../../../helpers/TestHelpers'

jest.mock('../../../../src/components/horizonCard/HorizonCardHeader', () => require('../../../mocks/HorizonCardHeader'))
jest.mock('../../../../src/components/horizonCard/HorizonCardGraph', () => require('../../../mocks/HorizonCardGraph'))
jest.mock('../../../../src/components/horizonCard/HorizonCardFooter', () => require('../../../mocks/HorizonCardFooter'))

expect.addSnapshotSerializer(new CustomSnapshotSerializer())

describe('HorizonCardContent', () => {
  describe('render', () => {
    beforeAll(() => {
      jest.spyOn(HelperFunctions, 'nothing').mockImplementation(() => html``)
    })

    it('sets dark mode when it is set to true on the config', async () => {
      const config: IHorizonCardConfig = {
        type: 'horizon-card',
        darkMode: true
      }

      const horizonCardContent = new HorizonCardContent(config, {} as THorizonCardData)
      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof horizonCardContent.render>
      element.templateResultFunction = () => horizonCardContent.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('does not set dark mode when it is set to false on the config', async () => {
      const config: IHorizonCardConfig = {
        type: 'horizon-card',
        darkMode: false
      }

      const horizonCardContent = new HorizonCardContent(config, {} as THorizonCardData)
      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof horizonCardContent.render>
      element.templateResultFunction = () => horizonCardContent.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('renders the header when show header returns true', async () => {
      const horizonCardContent = new HorizonCardContent({} as IHorizonCardConfig, {} as THorizonCardData)
      jest.spyOn((horizonCardContent as unknown as { showHeader: () => boolean }), 'showHeader').mockImplementation(() => true)

      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof horizonCardContent.render>
      element.templateResultFunction = () => horizonCardContent.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('does not render the header when show header returns false', async () => {
      const horizonCardContent = new HorizonCardContent({} as IHorizonCardConfig, {} as THorizonCardData)
      jest.spyOn((horizonCardContent as unknown as { showHeader: () => boolean }), 'showHeader').mockImplementation(() => false)

      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof horizonCardContent.render>
      element.templateResultFunction = () => horizonCardContent.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('renders the footer when show footer returns true', async () => {
      const horizonCardContent = new HorizonCardContent({} as IHorizonCardConfig, {} as THorizonCardData)
      jest.spyOn((horizonCardContent as unknown as { showFooter: () => boolean }), 'showFooter').mockImplementation(() => true)

      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof horizonCardContent.render>
      element.templateResultFunction = () => horizonCardContent.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })

    it('does not render the footer when show footer returns false', async () => {
      const horizonCardContent = new HorizonCardContent({} as IHorizonCardConfig, {} as THorizonCardData)
      jest.spyOn((horizonCardContent as unknown as { showFooter: () => boolean }), 'showFooter').mockImplementation(() => false)

      const element = window.document.createElement('test-element') as TemplateResultTestHelper<typeof horizonCardContent.render>
      element.templateResultFunction = () => horizonCardContent.render()
      window.document.body.appendChild(element)
      await element.updateComplete

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot()
    })
  })
})
