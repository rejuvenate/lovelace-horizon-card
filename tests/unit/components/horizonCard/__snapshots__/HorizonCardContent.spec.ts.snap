// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HorizonCardContent render does not render the footer when show footer returns false 1`] = `
<!---->
      <ha-card>
        <div class="horizon-card ">
          <!--?lit-->
      <div class="horizon-card-header">
        HORIZON CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="horizon-card-graph">
        HORIZON CARD GRAPH
      </div>
    
          <!--?lit--><!--?-->
        </div>
      </ha-card>
    
`;

exports[`HorizonCardContent render does not render the header when show header returns false 1`] = `
<!---->
      <ha-card>
        <div class="horizon-card ">
          <!--?lit--><!--?-->
          <!--?lit-->
      <div class="horizon-card-graph">
        HORIZON CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="horizon-card-footer">
        HORIZON CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;

exports[`HorizonCardContent render does not set dark mode when it is set to false on the config 1`] = `
<!---->
      <ha-card>
        <div class="horizon-card ">
          <!--?lit-->
      <div class="horizon-card-header">
        HORIZON CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="horizon-card-graph">
        HORIZON CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="horizon-card-footer">
        HORIZON CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;

exports[`HorizonCardContent render renders the footer when show footer returns true 1`] = `
<!---->
      <ha-card>
        <div class="horizon-card ">
          <!--?lit-->
      <div class="horizon-card-header">
        HORIZON CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="horizon-card-graph">
        HORIZON CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="horizon-card-footer">
        HORIZON CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;

exports[`HorizonCardContent render renders the header when show header returns true 1`] = `
<!---->
      <ha-card>
        <div class="horizon-card ">
          <!--?lit-->
      <div class="horizon-card-header">
        HORIZON CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="horizon-card-graph">
        HORIZON CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="horizon-card-footer">
        HORIZON CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;

exports[`HorizonCardContent render sets dark mode when it is set to true on the config 1`] = `
<!---->
      <ha-card>
        <div class="horizon-card horizon-card-dark">
          <!--?lit-->
      <div class="horizon-card-header">
        HORIZON CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="horizon-card-graph">
        HORIZON CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="horizon-card-footer">
        HORIZON CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;
